### PR TITLE
add support for revolt energy monitor nc-5462

### PR DESCRIPTION
**Pull request by nicolaihess**
Power Plug Energy Monitor NC-5642 (from www.revolt-power.de).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -315,8 +315,9 @@ See [CONTRIBUTING.md](./docs/CONTRIBUTING.md).
     [230]  Vauno EN8822C
     [231]  Govee Water Leak Detector H5054
     [232]  TFA Dostmann 14.1504.V2 Radio-controlled grill and meat thermometer
-    [233]* CED7000 Shot timer
+    [233]* CED7000 Shot Timer
```
Comment (zuckschwerdt) [inline README.md:318]: Can you rebase and `./maintainer_update.py` so this change does not appear?

File: README.md [hunk 2]
```diff
@@ -326,7 +327,7 @@ See [CONTRIBUTING.md](./docs/CONTRIBUTING.md).
   [-d <RTL-SDR USB device index>] (default: 0)
   [-d :<RTL-SDR USB device serial (can be set with rtl_eeprom -s)>]
 	To set gain for RTL-SDR use -g <gain> to set an overall gain in dB.
-	SoapySDR device driver is available.
+	SoapySDR device driver is not available.
```
Comment (zuckschwerdt) [inline README.md:330]: This change must be excluded from the PR

File: man/man1/rtl_433.1 [hunk 3]
```diff
@@ -163,7 +163,7 @@ RTL\-SDR device driver is available.
 To set gain for RTL\-SDR use \-g <gain> to set an overall gain in dB.
 .RE
 .RS
-SoapySDR device driver is available.
+SoapySDR device driver is not available.
```
Comment (zuckschwerdt) [inline man/man1/rtl_433.1:166]: This change must be excluded from the PR.

File: src/devices/revolt_nc5462.c [hunk 4]
```diff
@@ -0,0 +1,159 @@
+/** @file
+    Revolt NC5462 Energy Monitor
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:2]: Add a dot at the end.

File: src/devices/revolt_nc5462.c [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    Copyright (C) 2023 Nicolai Hess
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+*/
+/**
+  revolt energy power meter.
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:13]: Please remove the leading 2 spaces in *this* section.

File: src/devices/revolt_nc5462.c [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+   +----------+----------+
+
+
+
+  "Register" mode
+  (after pushing button on energy meter)
+  same 104 data pulses as in data mode, but first bit high and multiple rows of (the same)
+  data.
+
+  Pulses
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:54]: Review the formatting, this is markdown. Newlines don't actually break. Use list with dashes (`-`) for e.g. the items below.

File: src/devices/revolt_nc5462.c [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+  rtl_433 demodulation output
+  (normal data)
+  short_width: 200, long_width: 330, reset_limit: 240, sync_width: 10044
+  (detect flag)
+  short_width: 200, long_width: 330, reset_limit: 240, sync_width: 10044
+*/
+
+#include "decoder.h"
+#include "r_util.h"
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:68]: Do you really need r_util.h? You should not.

File: src/devices/revolt_nc5462.c [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+static int check_bitbuffer_row1byte12(bitbuffer_t *bitbuffer) {
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:70]: This should be placed into the decoder and not be a function.

File: src/devices/revolt_nc5462.c [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+      byte12 = bitbuffer->bb[0][11];
+      for (index=0; index<11;++index) {
+        checksum+=bitbuffer->bb[0][index];
+      }
+      return checksum == byte12;
+    }
+  return 0;
+}
+
+/**
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:86]: There must not be a (doc-)comment here! There is already above.

File: src/devices/revolt_nc5462.c [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+ decode
+*/
+static int revolt_nc5462_callback(r_device *decoder, bitbuffer_t *bitbuffer)
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:89]: Name it `revolt_nc5462_decode`.

File: src/devices/revolt_nc5462.c [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+{
+    bitrow_t *bb = bitbuffer->bb;
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:91]: If you work with only row 0 then it should be `uint8_t *b = bitbuffer->bb[0]` not `bb`.

File: src/devices/revolt_nc5462.c [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    uint16_t current;
+    uint8_t frequency;
+    uint16_t power;
+    uint8_t pf;
+    uint16_t energy;
+    uint8_t detect_flag;
+    bitbuffer_invert(bitbuffer);
+
+    if (check_bitbuffer_row1byte12(bitbuffer)) {
+      id = (bb[0][0]<<8) | (bb[0][1]);
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:104]: All indents are 4 spaces per level please.

File: src/devices/revolt_nc5462.c [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        detect_flag = 0;
+      }
+
+      voltage = bb[0][2];
+      current = bb[0][4] | bb[0][3]<<8;
+      frequency = bb[0][5];
+      power = bb[0][7] | bb[0][6]<<8;
+      pf = bb[0][8];
+      energy = bb[0][10] | bb[0][9]<<8;
+      data = data_make(
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:118]: Wrap in `    /* clang-format off */` and `    /* clang-format on */` please. Format as table, see other decoders.

File: src/devices/revolt_nc5462.c [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+                "frequency", "Frequency",  DATA_INT, frequency,
+                "power", "Power",  DATA_FORMAT, "%.02f W",DATA_DOUBLE, .1*power,
+                "power factor", "power factor",DATA_FORMAT, "%.02f Pf",  DATA_DOUBLE, .01*pf,
+                "energy", "energy",DATA_FORMAT, "%.02f kWh",  DATA_DOUBLE, .01*energy,
+                "detect flag", "Detect Flag", DATA_STRING, detect_flag ? "Yes" : "No",
+                NULL);
+      decoder_output_data(decoder, data);
+      return 1;
+    }
+    return 0;
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:132]: This return should be an early abort right at the top, not a nested block please.

File: src/devices/revolt_nc5462.c [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+static char* output_fields[] = {
+        "time",
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:136]: remove `time`

File: src/devices/revolt_nc5462.c [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+      current = bb[0][4] | bb[0][3]<<8;
+      frequency = bb[0][5];
+      power = bb[0][7] | bb[0][6]<<8;
+      pf = bb[0][8];
+      energy = bb[0][10] | bb[0][9]<<8;
+      data = data_make(
+                "model", "", DATA_STRING, "NC5462",
+                "id", "House Code",  DATA_INT, id,
+                "voltage", "Voltage",  DATA_INT, voltage,
+                "current", "Current ",DATA_FORMAT, "%.02f A ",  DATA_DOUBLE, .01 * current,
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:122]: usually we write `current * 0.01,`

File: src/devices/revolt_nc5462.c [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+                "frequency", "Frequency",  DATA_INT, frequency,
+                "power", "Power",  DATA_FORMAT, "%.02f W",DATA_DOUBLE, .1*power,
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:124]: all these keys should have units, e.g. `voltage_V`, `current_A`, `frequency_Hz`, `power_W`, `energy_kWh`.

File: src/devices/revolt_nc5462.c [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+                "power factor", "power factor",DATA_FORMAT, "%.02f Pf",  DATA_DOUBLE, .01*pf,
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:125]: no spaces in keys please, also below.

File: src/devices/revolt_nc5462.c [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+        "frequency",
+        "power",
+        "power factor",
+        "energy",
+        "detect flag",
+        NULL
+};
+
+
+r_device revolt_nc5462 = {
```
Comment (zuckschwerdt) [inline src/devices/revolt_nc5462.c:150]: Needs to be `r_device const revolt_nc5462 = {` -- since yesterday now ;)

**Review by zuckschwerdt — CHANGES_REQUESTED** [review]
Thanks, this looks good, some style changes though :)

File: src/devices/revolt_nc5462.c [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+      data = data_make(
+                "model", "", DATA_STRING, "NC5462",
+                "id", "House Code",  DATA_INT, id,
+                "voltage", "Voltage",  DATA_INT, voltage,
+                "current", "Current ",DATA_FORMAT, "%.02f A ",  DATA_DOUBLE, .01 * current,
+                "frequency", "Frequency",  DATA_INT, frequency,
+                "power", "Power",  DATA_FORMAT, "%.02f W",DATA_DOUBLE, .1*power,
+                "power factor", "power factor",DATA_FORMAT, "%.02f Pf",  DATA_DOUBLE, .01*pf,
+                "energy", "energy",DATA_FORMAT, "%.02f kWh",  DATA_DOUBLE, .01*energy,
+                "detect flag", "Detect Flag", DATA_STRING, detect_flag ? "Yes" : "No",
```
Comment (merbanan) [inline src/devices/revolt_nc5462.c:127]: Integrity tag is missing.

File: src/devices/revolt_nc5462.c [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+    uint8_t voltage;
+    uint16_t current;
+    uint8_t frequency;
+    uint16_t power;
+    uint8_t pf;
+    uint16_t energy;
+    uint8_t detect_flag;
+    bitbuffer_invert(bitbuffer);
+
+    if (check_bitbuffer_row1byte12(bitbuffer)) {
```
Comment (merbanan) [inline src/devices/revolt_nc5462.c:103]: The checksum check should return the proper type. Look at other decoders for inspiration.

**Comment by merbanan** [comment]
And send some signal samples if that has not been sent.

**Comment by nicolaihess** [comment]
Thanks for the review and feedback. I will try to apply the changes.
And I am about to add signal samples and device descriptions on rtl_433_tests
(Should have read the contribution page first :) )

**Review by zuckschwerdt — APPROVED** [review]

**Comment by zuckschwerdt** [comment]
Merge with some changes on top. Only noticeable changes are:
- "model" is now "Revolt-NC5462"
- "detect_mode" renamed to "button" with 0/1 output like in other decoders.

**Comment by nicolaihess** [comment]
thanks for finishing and taken the pull request.

**Pull request merged**